File: sync/sgl.py
# -*- coding: utf-8 -*-
"""Cliente de la API de SGL (sistema de gestión del depósito TML).

⚠ SOLO se usa el endpoint de LECTURA GetStock. Los endpoints GetDocumentos /
InsertDocumentos crean órdenes de trabajo y NO deben llamarse (los usa GBP).

El stock viene desglosado por estado; para comparar contra el stock físico
de GBP se usa el estado "FISICO" (validado contra datos reales: coincide
exacto en ~83% de los artículos; el resto son diferencias genuinas).
"""
import time
from datetime import datetime

import requests

import config
import db

_token_cache = {"token": None, "exp": 0}


def _token():
    if _token_cache["token"] and time.time() < _token_cache["exp"] - 60:
        return _token_cache["token"]
    r = requests.post(config.SGL_API_URL + "/api/Token", json={
        "clientId": config.SGL_CLIENT_ID,
        "clientSecret": config.SGL_CLIENT_SECRET,
        "grantType": "client_credentials",
    }, timeout=30)
    r.raise_for_status()
    tok = r.json()["access_token"]
    _token_cache["token"] = tok
    # el JWT dura bastante; renovamos cada 30 min por las dudas
    _token_cache["exp"] = time.time() + 1800
    return tok


def _num(v):
    try:
        return float(str(v).replace(",", "."))
    except (TypeError, ValueError):
        return 0.0


def fetch_stock():
    """Stock actual en SGL, agregado por SKU.

    Devuelve (filas, snapshot) donde filas = [(sku, fisico, descr)] y
    snapshot es la fecha/hora que informa SGL para la foto.
    """
    url = (f"{config.SGL_API_URL}/api/WMS/Stock/GetStock/"
           f"{int(config.SGL_CLIENTE)}/null/null/true")
    r = requests.get(url, headers={"Authorization": f"Bearer {_token()}"},
                     timeout=180)
    r.raise_for_status()
    data = r.json()
    if not isinstance(data, list):
        raise RuntimeError(f"Respuesta inesperada de SGL: {str(data)[:120]}")

    excl_pref = tuple(p.upper() for p in config.SGL_EXCLUDE_PREFIXES)
    excl_sku = {s.upper() for s in config.SGL_EXCLUDE_SKUS}

    por_sku = {}
    snapshot = None
    for row in data:
        sku = (row.get("articuloCodigo") or "").strip().upper()
        if not sku:
            continue
        # descartar ubicaciones/etiquetas logísticas (no son mercadería)
        if sku in excl_sku or sku.startswith(excl_pref):
            continue
        estado = (row.get("estadoDescripcion") or "").strip().upper()
        entry = por_sku.setdefault(sku, {"fisico": 0.0, "descr": ""})
        if estado == "FISICO":
            entry["fisico"] += _num(row.get("cantidad"))
        if not entry["descr"]:
            entry["descr"] = (row.get("articuloDescripcion") or "").strip()
        if not snapshot and row.get("fecha"):
            f, h = str(row["fecha"]), str(row.get("hora") or "")
            snapshot = (f"{f[:4]}-{f[4:6]}-{f[6:8]} {h}" if len(f) == 8 else f)
    rows = [(sku, v["fisico"], v["descr"]) for sku, v in por_sku.items()]
    return rows, snapshot


def sync_sgl(con):
    """Actualiza la tabla sgl_stock desde la API. Devuelve (filas, snapshot)."""
    rows, snapshot = fetch_stock()
    if not rows:
        raise RuntimeError("SGL devolvió 0 artículos")
    con.execute("DELETE FROM sgl_stock")
    con.executemany(db.upsert("sgl_stock", ("sku", "qty", "descr"), ("sku",)),
                    rows)
    con.commit()
    db.set_meta(con, "sgl_info", {
        "at": datetime.now().isoformat(timespec="seconds"),
        "file": None, "rows": len(rows),
        "source": "api", "snapshot": snapshot,
    })
    return len(rows), snapshot


if __name__ == "__main__":
    import sys
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8",
                                  errors="replace")
    con = db.connect()
    n, snap = sync_sgl(con)
    con.close()
    print(f"SGL actualizado: {n} SKUs (foto {snap})")

File: src/lib/panel/sgl.js
// Cliente de la API de SGL (sistema del depósito TML) — portado de sgl.py.
// ⚠ SOLO se usa el endpoint de LECTURA GetStock. NUNCA GetDocumentos /
// InsertDocumentos (crean órdenes de trabajo).

export const SGL_EXCLUDE_PREFIXES = ["PALLET", "XEROXALTPALL", "DEVOLUC"];
export const SGL_EXCLUDE_SKUS = ["PO", "NTB1", "PISTOLA TERMOMETRO"];

function cfg() {
  return {
    url: process.env.SGL_API_URL || "https://conexion.tmlogistica.com.ar/SGLAPIS_TML_PROD",
    clientId: process.env.SGL_CLIENT_ID || "argcol",
    clientSecret: process.env.SGL_CLIENT_SECRET || "",
    cliente: parseInt(process.env.SGL_CLIENTE || "54", 10),
  };
}

const numSgl = (v) => {
  const n = parseFloat(String(v ?? "0").replace(",", "."));
  return Number.isFinite(n) ? n : 0;
};

async function getToken() {
  const c = cfg();
  const r = await fetch(`${c.url}/api/Token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      clientId: c.clientId,
      clientSecret: c.clientSecret,
      grantType: "client_credentials",
    }),
  });
  if (!r.ok) throw new Error(`Token SGL HTTP ${r.status}`);
  const j = await r.json();
  if (!j.access_token) throw new Error("SGL no devolvió access_token");
  return j.access_token;
}

/** Stock actual en SGL agregado por SKU (estado FISICO). Devuelve {rows, snapshot}. */
export async function fetchStock() {
  const c = cfg();
  const token = await getToken();
  const url = `${c.url}/api/WMS/Stock/GetStock/${c.cliente}/null/null/true`;
  const r = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!r.ok) throw new Error(`GetStock HTTP ${r.status}`);
  const data = await r.json();
  if (!Array.isArray(data)) throw new Error("Respuesta inesperada de SGL");

  const exPref = SGL_EXCLUDE_PREFIXES.map((p) => p.toUpperCase());
  const exSku = new Set(SGL_EXCLUDE_SKUS.map((s) => s.toUpperCase()));
  const bySku = {};
  let snapshot = null;
  for (const row of data) {
    const sku = (row.articuloCodigo || "").trim().toUpperCase();
    if (!sku) continue;
    if (exSku.has(sku) || exPref.some((p) => sku.startsWith(p))) continue;
    const estado = (row.estadoDescripcion || "").trim().toUpperCase();
    const e = (bySku[sku] ||= { fisico: 0, descr: "" });
    if (estado === "FISICO") e.fisico += numSgl(row.cantidad);
    if (!e.descr) e.descr = (row.articuloDescripcion || "").trim();
    if (!snapshot && row.fecha) {
      const f = String(row.fecha);
      const h = String(row.hora || "");
      snapshot = f.length === 8 ? `${f.slice(0, 4)}-${f.slice(4, 6)}-${f.slice(6, 8)} ${h}` : f;
    }
  }
  const rows = Object.entries(bySku).map(([sku, v]) => ({ sku, qty: v.fisico, descr: v.descr }));
  return { rows, snapshot };
}

/** Reemplaza el contenido de sgl_stock y actualiza meta 'sgl_info'. */
export async function replaceSglStock(prisma, rows, sglInfo) {
  await prisma.$executeRawUnsafe(`DELETE FROM sgl_stock`);
  const CH = 500;
  for (let i = 0; i < rows.length; i += CH) {
    const chunk = rows.slice(i, i + CH);
    const vals = [];
    const params = [];
    chunk.forEach((r, j) => {
      const b = j * 3;
      vals.push(`($${b + 1}, $${b + 2}, $${b + 3})`);
      params.push(String(r.sku), Number(r.qty) || 0, r.descr || "");
    });
    await prisma.$executeRawUnsafe(
      `INSERT INTO sgl_stock (sku, qty, descr) VALUES ${vals.join(", ")} ` +
        `ON CONFLICT (sku) DO UPDATE SET qty = EXCLUDED.qty, descr = EXCLUDED.descr`,
      ...params
    );
  }
  await prisma.$executeRawUnsafe(
    `INSERT INTO meta (key, value) VALUES ('sgl_info', $1) ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value`,
    JSON.stringify(sglInfo)
  );
}
